docs/configuration/uds-operator.md

```diff
@@ -207,6 +207,48 @@ spec:
 
 This configuration does not create a secret in the cluster and instead tells the UDS Operator to create a public client (one that requires no auth secret) that enables the `oauth2.device.authorization.grant.enabled` flow and disables the standard redirect auth flow.  Because this creates a public client configuration that deviates from this is limited - if your application requires both the Device Authorization Grant and the standard flow this is currently not supported without creating two separate clients.
 
+### Creating a UDS Package with a Service Account Roles client
+
+Some applications may need to access resources / obtain OAuth tokens on behalf of *themselves* vice users. This may be needed to allow API access to Authservice protected applications (outside of a web browser). This is commonly used in machine-to-machine authentication for automated processes. This type of grant in OAuth 2.0 is known as the [Client Credentials Grant](https://oauth.net/2/grant-types/client-credentials/) and is supported in a UDS Package with the following configuration:
+
+```yaml
+apiVersion: uds.dev/v1alpha1
+kind: Package
+metadata:
+  name: client-cred
+  namespace: argo
+spec:
+  sso:
+    - name: httpbin-api-client
+      clientId: httpbin-api-client
+      standardFlowEnabled: false
+      serviceAccountsEnabled: true
+
+      # By default, Keycloak will not set the audience `aud` claim for service account access token JWTs.
+      # You can optionally add a protocolMapper to set the audience.
+      # If you map the audience to the same client used for authservice, you can enable access to authservice protected apps with a service account JWT.
+      protocolMappers:
+        - name: audience
+          protocol: "openid-connect"
+          protocolMapper: "oidc-audience-mapper"
+          config:
+            included.client.audience: "uds-core-httpbin" # Set this to match the app's authservice client id
```

Comment:
This is the one part of this setup that slightly worries me and I want to make sure I understand. By adding this protocol mapper to a `serviceAccountsEnabled` client, someone can effectively set the audience here to any client. Our current istio/authservice protections would not identify JWTs from this client as being any different from the "real" authservice client since we effectively just check the issuer and audience (which would be 1:1 with the authservice client). I think that is the intent here of course, effectively enabling machine auth on authservice protected apps.

Are there any concerns with this though, especially thinking about multi-tenant clusters? Tenant A could have an authservice protected app, and Tenant B could use a `serviceAccountsEnabled` client to access that app (I think?). Are there things we could/should recommend that the application do to prevent this? Or modifications we could make to our Istio resources to conditionally allow this type of behavior, only if the app is expecting machine auth?

I may also just be missing things here but want to make sure I understand the intent with this flow and any risks/mitigations that are or could be present.

Reply:
What you're describing is exactly the use-case.  I want to use the `client_id` (or a group / role etc associated with the client) as an authZ key to solve a multi-tenancy issue.  BUT, all of the different tenants need to use the same API.  I originally had this set up so the API did the JWT verification and would accept multiple "audiences" but we always had a goal of getting authservice to take over this role.

We looked at manually supplying the `AuthorizationPolicy` and `RequestAuthentication` CRs in the chart for my project to allow multiple audiences, but figured it's probably safer to let the operator handle this.

I think this is the most intuitive way:  The API is the "audience", which makes sense, but the machines that access the API are doing so on behalf of different companies and development teams, hence the different clients that have different permissions.

Reply:
> Are there any concerns with this though, especially thinking about multi-tenant clusters?
We already don't do any real multi tenant isolation at configuration time, the person deploying a Package CR is basically assumed to be cluster admin I think? Or is that the place of Exemptions, only those are privileged configuration? We already don't isolate Keycloak clients: it turns out two Package CRs can reference the same Keycloak client, and pepr will fetch it out of the store. You can also do other things to hurt multi-tenant isolation, like add any NetworkPolicy you want.

We could consider each custom protocolMappers to be an exemption and require documentation? This is really why I was recommending an allowlist from the beginning for this feature, with the right config you can make your client do anything (probably even takeover Keycloak by overriding claims). 

(I really would like to sketch out a threat model and what implied permissions are associated with each CR)

I'm doing a little bit more testing, but as far as I can tell a Keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`. I also think this audience mapper should work with Device Flow, is that maybe worth documenting too?

Reply:
Also, I think this isn't intuitive, but a `serviceAccount` type client does not automatically come with an `audience` field in keycloak.  Since a lot of people use `aud` to do validation we needed to test a code pattern to add the field.

Reply:
Yeah sorry @ldgriswold was tracking that the intent of these clients was exactly as described - authenticating using one client against a different authservice app (one that is protected by a different client). My concern was more around potential for two app teams to be working in the same cluster and one to maliciously gain access to the others application using a service account client. Talked async with @bburky about this a bit and I think it's probably not worth addressing at this point since there are other multi-tenant considerations like this that we haven't addressed and don't have a great story around how we want this to look. For now I think we can make the assumption that `Package` CR resources are likely managed by a cluster administrator and/or created via "approved" UDS packages that have been vetted. There might be future things to evaluate with this like restrictions around the audience protocol mapper or service account clients to better ensure they are "cluster admin" functions.

I do think it may be worth adding a quick note around what @bburky commented above...
> a Keycloak service account JWT can be distinguished by a username prefix of `service-account-` and a new claim called `client_id`

Not sure we need to document the device flow + service account piece at this point though.